### DISCUSSION TITLE
Storing DateTime in Milliseconds UTC

### DISCUSSION THREAD

**Discussion by FaFre** [simolus3/drift#2403] [General]
At the moment Drift is offering two options to store `DateTime`s: ISO-String and Unix Timestamp (in Seconds).

I want to store `DateTime` in milliseconds to be able to do apply some math during my select. Also the `DateTime` should be stored in UTC when writing, and automatically converted back to local time when reading.

Is there some way to override the standard behavior of DateTime conversion in Drift?

**Comment by kuhnroyal** [comment]
You can use https://drift.simonbinder.eu/docs/advanced-features/type_converters/

**Reply by simolus3** [reply]
To add to this answer: Drift exposes date and time functions builtin to sqlite3 as part of its Dart API. So every `Expression<DateTime>` needs to be in a format that these functions can work with. And since sqlite3 only supports a precision of seconds for those APIs, that's all we can generally support in drift as well.

When date times are stored as text, drift will actually keep the milliseconds part, so no information is lost by storing values in that format. The precision is only lost when invoking datetime functions from sqlite3 since they don't return milliseconds in their format. We can't do the same thing when storing these values as unix timestamps since we don't know whether a given number is stored in millisecond accuracy or in seconds. With `2023-04-30T14:09:47.367590 +02:00` it's obvious, but `1682856587` could refer to the same time or `1970-01-20T12:27:36.587` depending on the way it's interpreted. So the whole thing would be more dangerous, and no option exists to store them as milliseconds.

So maybe the question is: Can you math be expressed with datetime functions supported by sqlite3? In that case you could use the default format. If the transformation on text is to expensive or generally doesn't work, type converters are probably the only option. We can't support a custom behavior of mapping DateTimes to SQL either, since that format would have to be understood by the Dart API for datetime conversions too.

**Reply by FaFre** [reply]
Thanks a lot you two for your suggestions! I read the sqlite documentation about DateTime and figured that what I am doing is non-standard and I should use my own type converter for that. I have an append only table where I perform group by's, as well as range scans with relative time offsets (indexed as well as non-indexed). Since the table could grow up to several thousand rows using an integer fells better, although the ISO string wouldn't probably perform much worse.
